### PR TITLE
8240256: Better resource cleaning for SunPKCS11 Provider

### PR DESCRIPTION
**Pull request by coffeys**
Added capability to allow the PKCS11 Token to be destroyed once a session is logged out from. New configuration properties via pkcs11 config file. Cleaned up the native resource poller also.

New unit test case to test behaviour. Some PKCS11 tests refactored to allow pkcs11 provider to be configured (and tested) with a config file of choice.

Reviewer request @valeriepeng

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8240256](https://bugs.openjdk.java.net/browse/JDK-8240256): Better resource cleaning for SunPKCS11 Provider

### Reviewers
 * [Valerie Peng](https://openjdk.java.net/census#valeriep) (@valeriepeng - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3544/head:pull/3544` \
`$ git checkout pull/3544`

Update a local copy of the PR: \
`$ git checkout pull/3544` \
`$ git pull https://git.openjdk.java.net/jdk pull/3544/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3544`

View PR using the GUI difftool: \
`$ git pr show -t 3544`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3544.diff">https://git.openjdk.java.net/jdk/pull/3544.diff</a>

</details>

### CODE REVIEW

**Comment by valeriepeng** [comment]
I will take a look. Thanks!

**Review by valeriepeng — COMMENTED** [review]
Here are some comments. Mostly just nit. Will continue looking at the test changes. 
Thanks~ 
Valerie

File: test/jdk/sun/security/pkcs11/PKCS11Test.java [hunk 1]
```diff
@@ -536,36 +539,50 @@ public static void useNSS() {
...
+
+    // Run NSS testing on a Provider p configured with test nss config
+    public static void testNSS(PKCS11Test test, Provider p) throws Exception {
+        String nssConfig = getNssConfig();
+        if (nssConfig == null) {
+            // issue loading libraries
+            return;
+        }
+        p = getSunPKCS11(nssConfig);
+        test.premain(p);
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/PKCS11Test.java:555]: It seems a bit strange that the 2nd argument p isn't really used, but rather overridden with getSunPKCS11(String config) call. Do you mean to call getSunPKCS11(nssConfig, p) here if p is non-null?

File: test/jdk/sun/security/pkcs11/Provider/Login.java [hunk 2]
```diff
@@ -22,20 +22,33 @@
...
+
+        pkcs11Provider.logout();
+        WeakReference<SunPKCS11> weakRef = new WeakReference<>(pkcs11Provider);
+        pkcs11Provider = null;
+        for (int i = 0; i > 100; i++) {
+            System.gc();
+            Thread.sleep(100);
+        }
+        System.out.println("Finish: "+ weakRef.refersTo(null));
     }
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/Login.java:52]: What is this particular block of code for? main(Provider) is non-static and may be called more than once, but then you store the passed-in Provider object into a static field? Also, must the static field use type SunPKCS11? It seems AuthProvider is sufficient and most of the "ap->pkcs11Provider" changes can be avoided?

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
+/*
+ * Copyright (c) 2003, 2016, Oracle and/or its affiliates. All rights reserved.
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java:2]: New file, should just be 2021?

File: test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+import java.io.IOException;
+import java.lang.ref.WeakReference;
+import java.security.AuthProvider;
+import java.security.KeyStore;
+import java.security.Provider;
+import java.security.Security;
+
+import jdk.test.lib.util.ForceGC;
+
+public class MultipleLogins extends PKCS11Test {
```
Comment (valeriepeng) [inline test/jdk/sun/security/pkcs11/Provider/MultipleLogins.java:42]: The overall pattern seems a bit different from the typical PKCS11Test usage pattern. When a class extends PKCS11Test, the main test functionality is under its main(Provider) method. It is expected that the PKCS11Test class will call the main(Provider) method with the to-be-tested PKCS11 provider obj. Here it seems that you are doing that in the static main(String[]) method. Do you really need to extend PKCS11Test then?

**Comment by coffeys** [comment]
Thanks for the review @valeriepeng  Apologies for delay, had to finish up other tasks -- I've made all suggested edits that you made for src/ code and pushed those changes to my branch.

For the testing, then I recall the PKCS11Test framework not being able to return multiple PKCS11 Providers to me for testing (an array of providers) -- I thought editing the framework for such a use case would be useful but perhaps I'm better off just writing a customized test. I'm going to take a stab at that and will re-submit a test edit.

Apologies for including the Login.java/ Login.sh edits - they were old left over edits from testing efforts before moving on to develop the new MultipleLogins.java test.

**Comment by valeriepeng** [comment]
The update changes look good. Will wait for your test edit then.

**Review by valeriepeng — APPROVED** [review]
Changes look fine. Thanks, Valerie

**Pull request closed without merging**